### DISCUSSION TITLE
How to set a variable field of "Yield Stress" in the J2 plasticity model

### DISCUSSION THREAD

**Discussion by sSajjad90** [idaholab/moose#21261] [Q&A Modules: Solid mechanics] [answered]
Hi everyone, 
In a phase calculation, the **yield stress** is calculated in material and is **not** constant. Now, I am going to add this yield stress to the J2 plasticity model, although available userobjects, `TensorMechanicsHardeningConstant.C` for instance, take constant values.

```
[Materials]
  [./Hardening]
    type = PrecipitationHardening

    Delta_Sigma_p = 'Sigma_p'          //Declared Yeild Stress

    Time_step = 'time_step'
    BurgursVec = 'b_V'
    ShearModule = 'Mu'
    PoisionRatio = 'Nu'
    TaylorFactor = 'M_Taylor'
    Gama = 'Gama_TDP'
    PhaseNumber = 3
    q_exponent = 2
  [../]

[UserObjects]
  [./str]
    type = TensorMechanicsHardeningConstant

    value =         **//  hope to use 'Sigma_p' (material property) instead  of constant value**
  [../]
  [./j2]
    type = TensorMechanicsPlasticJ2
    yield_strength = str
    yield_function_tolerance = 1E-3
    internal_constraint_tolerance = 1E-9
  [../]
[]
```

In this way, I tried to change the type of parameter inside the `TensorMechanicsHardeningConstant.C`  and ` ***.h` files and got errors. (I tried `getMaterialProperty<Real>` and `adCoupledValue` and I got error)

This problem has already been discussed in #19454, but I could not realize how to figure it out. 

Would you please let me know the proper parameters to use the mentioned 'Sigma_p'  from the material calculation in the userobject?

Sincerely,
Sajjad

**Comment by GiudGiud** [comment, marked as answer]
so you want to use a material property?
Or a variable? or a function?

There are plenty of examples for using each of these.

**Reply by sSajjad90** [reply]
I am going to use a material property which is not constant and would be changed via time

**Reply by GiudGiud** [reply]
You can have a look at ADMatDiffusion for how to work with those

**Reply by sSajjad90** [reply]
Please look at the below codes, derived from `TensorMechanicsHardeningModel`, providing yield_strength for the `TensorMechanicsPlasticJ2` userobject.
I understand that `_qp` is defined for regular materials using the computeQpProperties routine. But what exactly I want is that material property that is calculated each time (and will update next time). Now, I am wondering how I could use this material property in the userobject.

```
#include "YieldStressofBainaryPrecipitationHardening.h"

registerMooseObject("microstructureApp", YieldStressofBainaryPrecipitationHardening);

InputParameters
YieldStressofBainaryPrecipitationHardening::validParams()
{
  InputParameters params = TensorMechanicsHardeningModel::validParams();
  params.addRequiredParam<MaterialPropertyName>("value_0", "The yield strength when internal variable = 0");

  params.addClassDescription("Hardening defined by Percipitation");

  return params;
}

YieldStressofBainaryPrecipitationHardening::YieldStressofBainaryPrecipitationHardening(
    const InputParameters & parameters)
  : TensorMechanicsHardeningModel(parameters),
    _value_0(getMaterialProperty<Real>("value_0"))
{
}
Real
YieldStressofBainaryPrecipitationHardening::value(Real intnl) const
{
  return _value_0[_qp];
}
Real
YieldStressofBainaryPrecipitationHardening::derivative(Real intnl)  const
{
  return 0;
}
std::string
YieldStressofBainaryPrecipitationHardening::modelName() const
{
  return "PrecipitationHardening";
}
```

```
#pragma once

#include "TensorMechanicsHardeningModel.h"


class YieldStressofBainaryPrecipitationHardening : public TensorMechanicsHardeningModel
{
public:
  static InputParameters validParams();

  YieldStressofBainaryPrecipitationHardening(const InputParameters & parameters);

  virtual Real value(Real intnl) const override;

  virtual Real derivative(Real intnl) const override;

  virtual std::string modelName() const override;


private:
  const MaterialProperty<Real> & _value_0;
};

```


And in the assembly file, `Sigma_p` is calculated in material properties and will be used in the `./str` useobject at each time step.


```
[Materials]
  [./Hardening]
    type = PrecipitationHardening

    Delta_Sigma_p = 'Sigma_p'          //Declared Yeild Stress

    Time_step = 'time_step'
    BurgursVec = 'b_V'
    ShearModule = 'Mu'
    PoisionRatio = 'Nu'
    TaylorFactor = 'M_Taylor'
    Gama = 'Gama_TDP'
    PhaseNumber = 3
    q_exponent = 2
  [../]
[]

[UserObjects]
  [./str]
    type = YieldStressofBainaryPrecipitationHardening
    value = Sigma_p        
  [../]
  [./j2]
    type = TensorMechanicsPlasticJ2
    yield_strength = str
    yield_function_tolerance = 1E-3
    internal_constraint_tolerance = 1E-9
  [../]
[]

```

**Reply by GiudGiud** [reply]
Hello

Material properties are updated on every linear iteration, so as often as possible, so you do not need to worry about when it is updated. 

I'm not sure about your implementation though, it seems you are not replacing what was previously a constant with a material property?

Guillaume

**Reply by sSajjad90** [reply]
Hi, 
I tried to replace the constant value taking in the `TensorMechanicsHardeningConstant` with a material property, but it does not work.

Let me explain my problem with a simple example. Imagine yield strengthening depends on `time_step` and calculated by a `ParsedMaterial`, which changes based on the time `t`, like below. 
This `Yiled_Stress` must be used in `TensorMechanicsPlasticJ2`. For this matter, another `userobject` is required to get the value from the material block and provide it for the plasticity model. (`[./str]` block)
Available userobjects that provide yield_strength like `TensorMechanicsHardeningConstant`, `TensorMechanicsHardeningPowerRule`, etc., can not get this type of property. I tried rewriting the `TensorMechanicsHardeningPowerRule` code to get material property; however, it does not work.
My problem is to define a proper useruoject in the `[./str]` block to get the Yiled_Stress.
```
[Materials]
 [./Hardening]
   type = ParsedMaterial
   f_name = 'Yiled_Stress'
   material_property_names =  '75 + t * 0.3'
   function = 't'
 [../]
[]

[UserObjects]
  [./str]
    type = 
    value = 'Yiled_Stress'        
  [../]
  [./j2]
    type = TensorMechanicsPlasticJ2
    yield_strength = str
    yield_function_tolerance = 1E-3
    internal_constraint_tolerance = 1E-9
  [../]
[]
```


Many thanks for your kind support,

**Reply by GiudGiud** [reply]
What do you mean by "it does not work" ?
Does it not compile?
Is it not retrieving the right value?
Is it not converging?

**Reply by sSajjad90** [reply]
It did not compile.  And the error is in the picture for the following  `YieldStressofBainaryPrecipitationHardening` userobject.

![image](https://user-images.githubusercontent.com/78385504/173245034-b49f3dd1-607e-4f59-8c85-5f8ec18b1001.png)

```

#include "YieldStressofBainaryPrecipitationHardening.h"

registerMooseObject("microstructureApp", YieldStressofBainaryPrecipitationHardening);

InputParameters
YieldStressofBainaryPrecipitationHardening::validParams()
{
  InputParameters params = TensorMechanicsHardeningModel::validParams();
  params.addRequiredParam<MaterialPropertyName>("value_0", "The yield strength when internal variable = 0");

  params.addClassDescription("Hardening defined by Percipitation");

  return params;
}

YieldStressofBainaryPrecipitationHardening::YieldStressofBainaryPrecipitationHardening(
    const InputParameters & parameters)
  : TensorMechanicsHardeningModel(parameters),
    _value_0(getMaterialProperty<Real>("value_0"))
{
}
Real
YieldStressofBainaryPrecipitationHardening::value(Real intnl) const
{
  return _value_0[_qp];
}
Real
YieldStressofBainaryPrecipitationHardening::derivative(Real intnl)  const
{
  return 0;
}
std::string
YieldStressofBainaryPrecipitationHardening::modelName() const
{
  return "PrecipitationHardening";
}

```

```
#pragma once

#include "TensorMechanicsHardeningModel.h"


class YieldStressofBainaryPrecipitationHardening : public TensorMechanicsHardeningModel
{
public:
  static InputParameters validParams();

  YieldStressofBainaryPrecipitationHardening(const InputParameters & parameters);

  virtual Real value(Real intnl) const override;

  virtual Real derivative(Real intnl) const override;

  virtual std::string modelName() const override;


private:
  const MaterialProperty<Real> & _value_0;
};
```

**Reply by GiudGiud** [reply]
So is this code trying to 
- define the material property you want to use
OR
- use that material property in the tensor mechanics model

Seems to me like there is a confusion here

**Reply by sSajjad90** [reply]
Sorry for my bad explanation.

use that material property in the tensor mechanics model

**Reply by GiudGiud** [reply]
Ok I dived a little bit into this code and it doesnt look like it is set up to use a material at all.

Do you need both spatial and time dependence? Or just time dependence?
If time dependence only you should use a Function.

**Reply by sSajjad90** [reply]
Actually, I need both spatial and time dependence

**Reply by GiudGiud** [reply]
Then I would modify the code that **uses** the TensorMechanicsHardeningModel to either:
- use a material property instead. It's a much larger rework.
- pass a positional argument instead of just `int intnl` down to the TensorMechanicsHardeningModels.

**Reply by sSajjad90** [reply]
Do you mean to change it like below?

It compiled and I also could get some results for a simulation.


Many thanks for your consideration and support.

```
#include "YieldStressofBainaryPrecipitationHardening.h"

registerMooseObject("microstructureApp", YieldStressofBainaryPrecipitationHardening);

InputParameters
YieldStressofBainaryPrecipitationHardening::validParams()
{
  InputParameters params = TensorMechanicsHardeningModel::validParams();
  params.addRequiredParam<std::string>("value_0", "The yield strength when internal variable = 0");



  params.addClassDescription("Hardening defined by Percipitation");

  return params;
}

YieldStressofBainaryPrecipitationHardening::YieldStressofBainaryPrecipitationHardening(
    const InputParameters & parameters)
  : TensorMechanicsHardeningModel(parameters),
    _value_0(getMaterialProperty<Real>(getParam<std::string>("value_0")))


{
}

Real
YieldStressofBainaryPrecipitationHardening::value(Real qp) const
{
  return _value_0[qp];

}

Real
YieldStressofBainaryPrecipitationHardening::derivative(Real qp)  const
{
  return 0;
}

std::string
YieldStressofBainaryPrecipitationHardening::modelName() const
{
  return "PrecipitationHardening";
}

```

**Reply by GiudGiud** [reply]
qp is an integer, it's an index, not a Real.
This looks problematic really. Are you setting the qp to be the quadrature index?

_value_0[_qp] is only equal to what you want when the material property has been properly re-initialized on the element. To ensure this is true it needs to be adequately called in an object that has a loop over quadrature points (for example a kernel or a material property). This generalUO is not performing this, I suspect _value_0 does not get updated at all

**Reply by sSajjad90** [reply]
You are right.  The results are bizarre. 

So, as I understood, this kind of property `_value_0[_qp]` can not be used in this userobject. 

Do you have any suggestions that I could use this particular material property (as the yield_strength ) in plasticity modeling?


To use a material property instead of userobject for the plasticity model, is there any alternative I could use with some changes for the yield_strength?

**Reply by GiudGiud** [reply]
You have to change the code that is calling these stress hardening models.
Also when you say you need spatial dependence, how coarse is that dependence going to be?
